[#15354] Component initialization tracker

jabolina opened the pull request:
* Track the time between different stages during initialization.
* Only tracks if trace is enabled.

I am using atomic long, but maybe it is not needed.

Closes #15354.

wburns commented on core/src/main/java/org/infinispan/factories/impl/ComponentRegistryTimeTracker.java:
```diff
@@ -0,0 +1,224 @@
+   private static final class Initialization implements Comparable<Initialization> {
+      private static final long NOT_INITIALIZED = 0L;
+      private final String name;
+      private final AtomicLong instantiating = new AtomicLong(NOT_INITIALIZED);
+      private final AtomicLong instantiated = new AtomicLong(NOT_INITIALIZED);
+      private final AtomicLong wiring = new AtomicLong(NOT_INITIALIZED);
+      private final AtomicLong wired = new AtomicLong(NOT_INITIALIZED);
+      private final AtomicLong starting = new AtomicLong(NOT_INITIALIZED);
+      private final AtomicLong started = new AtomicLong(NOT_INITIALIZED);
```
We can replace these with AtomicFieldUpdaters as well to reduce the overhead of all of these.

wburns commented on core/src/main/java/org/infinispan/factories/impl/ComponentRegistryTimeTracker.java:
```diff
@@ -0,0 +1,224 @@
+      }
+
+      public void instantiating(long value) {
+         instantiating.compareAndSet(NOT_INITIALIZED, value);
```
Should we error if this was set to something other than not initialized?

jabolina replied:
Makes sense. And could help identify something going wrong in the initialization.

wburns commented on core/src/main/java/org/infinispan/factories/impl/ComponentRegistryTimeTracker.java:
```diff
@@ -0,0 +1,224 @@
+
+   ComponentRegistryTimeTracker(BasicComponentRegistry registry, boolean global) {
+      this.registry = registry;
+      this.timeService = global ? null : registry.getComponent(TimeService.class).wired();
```
What about if a registry is restarted, don't we need to empty these values out?

wburns commented on core/src/main/java/org/infinispan/factories/impl/ComponentRegistryTimeTracker.java:
```diff
@@ -0,0 +1,224 @@
+   @Override
+   public String dump() {
+      StringBuilder sb = new StringBuilder();
+      List<Initialization> c = components.values().stream()
```
Should we not clear this map after it has completed? Or at least let the GC collect this object somehow.

wburns replied:
Maybe we want to change the class(es) that reference this to remove references to it.

jabolina replied:
Yeah, I was thinking about that yesterday. I guess removing the reference is easier, so it also works as a reset :)

wburns commented on core/src/main/java/org/infinispan/factories/impl/BasicComponentRegistryImpl.java:
```diff
@@ -67,6 +71,14 @@ public <T> ComponentRef<T> lazyGetComponent(Class<T> componentType) {
+   @Override
+   public void blameInitialization() {
+      if (tracker != null && log.isTraceEnabled()) {
+         log.tracef("Initialization blame of %s:%n%s", scope, tracker.dump());
```
I probably wouldn't say "blame" but maybe something like "Component initialization performance:" or something like that.

wburns commented on core/src/main/java/org/infinispan/factories/impl/ComponentRegistryTimeTracker.java:
```diff
@@ -0,0 +1,236 @@
+      List<Initialization> c = components.values().stream()
+            .sorted(Comparator.reverseOrder())
+            .toList();
+      for (Initialization initialization : c) {
```
I wonder if we should condense the components that took < 1 ms to not spam the log as much as well? Wdyt?

wburns commented on core/src/main/java/org/infinispan/factories/impl/ComponentRegistryTimeTracker.java:
```diff
@@ -0,0 +1,236 @@
+   }
+
+   private static final class Initialization implements Comparable<Initialization> {
+      private static final long NOT_INITIALIZED = 0L;
```
I wonder if this should be -1 so we can detect if a component was not tracked or a double update? Otherwise I can't tell the difference between really fast and not captured.

wburns reviewed (commented):
I noticed the timing doesn't work (for obvious reasons) on the TimeService object

```
org.infinispan.commons.time.TimeService:
	-> instantiating: 0 ms
	->  instantiated: 0 ms
	->        wiring: 0 ms
	->         wired: 1754596670672 ms
	->      starting: 0 ms
	->         total: 1754596670672 ms
```

jabolina commented:
Applied the fixes. The problem with `TimeService` was because the instantiating time wasn't recorded. I've fixed to only dump valid components. I also fixed to remove the component when it is replaced because it will go through the states again with the same name.

wburns commented:
Integrated into main, thanks @jabolina !
